src/include/sof/lib/ams.h

```diff
@@ -0,0 +1,340 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * Author: Krzysztof Frydryk <krzysztofx.frydryk@intel.com>
+ *
+ */
+
+#ifndef AM_SERVICE_H
+#define AM_SERVICE_H
+
+#include <errno.h>
+#include <sof/schedule/task.h>
+#include <ipc/topology.h>
+#include <rtos/alloc.h>
+#include <sof/coherent.h>
+#include <sof/lib/uuid.h>
+
+#define AMS_INVALID_MSG_TYPE 0
+#define AMS_SERVICE_UUID_TABLE_SIZE 49
+#define AMS_INVALID_SLOT 0xFF
+#define AMS_ANY 0xFFFF
+
+#define AMS_ROUTING_TABLE_SIZE 41
+#define AMS_MAX_MSG_SIZE 0x1000
+
+#define AMS_SLOT_SIZE(msg) (AMS_MESSAGE_SIZE(msg) + sizeof(uint16_t) * 2)
+#define AMS_MESSAGE_SIZE(msg) (sizeof(*msg) - sizeof(char) + (sizeof(char) * (msg->message_length)))
+
+/**
+ * \brief IXC message payload
+ *
+ * ams_message_payload - contains the actual Async Msg payload
+ */
+struct ams_message_payload {
+	uint32_t message_type_id;
+	uint16_t producer_module_id;
+	uint16_t producer_instance_id;
+	uint32_t message_length;
+	uint8_t *message;
+};
+
+struct ams_slot {
+	uint16_t module_id;
+	uint16_t instance_id;
+	union {
+		struct ams_message_payload msg;
+		uint8_t msg_raw[AMS_MAX_MSG_SIZE];
+	} u;
+	uint32_t __aligned(PLATFORM_DCACHE_ALIGN) pad[0];
+};
+
+/**
+ * \brief ams_msg_callback_fn
+ *
+ * Each subscriber provides this handler function for each message ID
+ */
+typedef void (*ams_msg_callback_fn)(const struct ams_message_payload *const ams_message_payload,
+				    void *ctx);
+
+
+/**
+ * \brief Internal struct ams_consumer_entry
+ *
+ * Describes a single consumer's subscription to a single message.
+ * Array of 'ams_consumer_entry' structs forms AsyncMessageService's routing
+ * table which allows for message dispatch.
+ */
+struct ams_consumer_entry {
+	/* Message ID that will be routed via this entry */
+	uint32_t message_type_id;
+	/* Callback provided by the subscribed consumer */
+	ams_msg_callback_fn consumer_callback;
+	/* Additional context for consumer_callback (optional) */
+	void *ctx;
```

Comment:
context provided by the source? Or context set by the consumer for convenience (i.e. a private_data field?)

Reply:
ok